### ISSUE TITLE
Cloned `TraitTypes` don't properly account for `callable_and_args` defaults

### ISSUE THREAD

**Issue by corranwebster** [enthought/traits#1629]
The following code segfaults:
``` python
from traits.api import TraitType, DefaultValue, HasTraits

class Foo:
   def __init__(self, value):
       self.value = value


class FooTrait(TraitType):

    default_value_type = DefaultValue.callable_and_args

    def __init__(self, default_value=None, **metadata):
        default_value = self.validate(None, None, default_value)
        default_value = (Foo, (default_value.value,), {})
        super().__init__(default_value, **metadata)

    def validate(self, object, name, value):
        if isinstance(value, Foo):
            return value
        else:
            return Foo(value)
    
# This is fine and expected
default = Foo(10)

class FooWrapper(HasTraits):
    foo = FooTrait(default)

foo_wrapper = FooWrapper()
assert foo_wrapper.foo.value == 10
assert foo_wrapper.foo is not default

# But this sort of thing is common
my_foo_trait = FooTrait(20)

class FooWrapper2(HasTraits):
    # clone the trait and override the default
    foo = my_foo_trait(10)

# note that this is deeply suspicious
print(my_foo_trait(10).default_value())

foo_wrapper2 = FooWrapper2()
# segfault
print(foo_wrapper2.foo.value)
```

Because clone's don't call `__init__` the new default value is not updated to the expected `(callable, args, kwargs)` form for a "callable and args"-style default, and is instead simply validated here: https://github.com/enthought/traits/blob/d1949c6e8bd2a22fa81d97a2dea10c51caf019b6/traits/trait_type.py#L307-L315
So the default value on the CTrait (as seen in the example) is `(7, <Foo instance>)`.  When the default value is requested, this then segfaults.

Note that at least part of this is that the default generation code for callable and args assumes that the user hasn't made an error in setting up the form of the default value correctly and will segfault even without the cloning if, for example, you do something like `default_value = (Foo, (default_value.value), {})` (note the missing comma in the tuple) as I found out while preparing the example above!

Work-around for my needs is to override `__call__` or `clone` to do the right processing on the defaults.

A secondary thing that will need to be done is an audit to see what existing `TraitType` subclasses in ETS exist that may be vulnerable to this.  I am suspicious of `BaseInstance` and the `PyfaceFont` trait at a minimum.

**Comment by corranwebster** [comment]
Confirming that a work-around is something like:
``` python
class FooTrait(TraitType):

    default_value_type = DefaultValue.callable_and_args

    def __init__(self, default_value=None, **metadata):
        default_value = self.validate(None, None, default_value)
        default_value = (Foo, (default_value.value,), {})
        super().__init__(default_value, **metadata)

    def clone(self, default_value=NoDefaultSpecified, **metadata):
        new = super().clone(NoDefaultSpecified, **metadata)
        if default_value is not NoDefaultSpecified:
            default_value = self.validate(None, None, default_value)
            new.default_value = (Foo, (default_value.value,), {})
        return new

    def validate(self, object, name, value):
        if isinstance(value, Foo):
            return value
        else:
            return Foo(value)
```

**Comment by mdickinson** [comment]
I think I have a solution for this. PR soon.

**Comment by mdickinson** [comment]
There's a design question hidden here, to do with subclassing and `Instance` traits. Suppose that we have:

```python
class HasRange(HasTraits):
    foo = Instance(range)
```

Then which of the following should be supported? Either:

```python
class HasRangeSpecialCase(HasTraits):
    foo = range(42)
```
where the intended behaviour is that we supply the specific range `range(42)` as the default, or:
```python
class HasRangeSpecialCase(HasTraits):
    foo = 42
```
where the intention is that `42` is an _argument_ to the `range` constructor?

I don't think we can support both. My inclination is to go with supporting the first but not the second.

**Comment by mdickinson** [comment]
(Quite separately, I'm inclined to think that `Instance` should grow the ability to specify a default value directly, rather than forcing use of `args` and `kw`.)

**Comment by mdickinson** [comment]
#601 is related

**Comment by mdickinson** [comment]
Fixed in #1631. There are still some serious issues with trait inheritance (see #1630 for example), but fixing those is for another day.
